### ISSUE TITLE
Ensure this works for `RegExp` subclasses

### ISSUE THREAD

**Issue by ljharb** [mikesamuel/regexp-make-js#11]
I don't see spec text yet, so it may not be a problem yet - but let's make sure that 

```js
class Foo extends RegExp {} ; Foo.make`${foo}`
```

returns an `instanceof Foo`, not just `RegExp`.

**Comment by mikesamuel** [comment]
So we would use the `this` value as the constructor?

Would it be preferable to invoke the constructor assuming it takes (source, flags) or to try to do something Object.create based and then call the RegExp constructor to initialize if possible?

**Comment by bergus** [comment]
Just use [`Reflect.construct`](http://www.ecma-international.org/ecma-262/6.0/#sec-reflect.construct). I think it's safe to assume it takes (source, flags). Using [RegExpInitialize](http://www.ecma-international.org/ecma-262/6.0/#sec-regexpinitialize) directly should be avoided.

**Comment by ljharb** [comment]
It's the same logic as `Array.from`/`Array.of` imo - it uses the `this` value for the species-or-constructor.

Why would I want to create a `RegExp` and then use `Reflect.construct` in order to get a subclass of `RegExp`, when I only want the subclass already? Wouldn't that call `RegExpInitialize` twice when I only need it once?

**Comment by bergus** [comment]
Yeah, just like `Array.of`/`Array.from`/`Promise.resolve` - not using `@@species` though, that's for instance methods.

So `new this(pattern, flags)` (or `Reflect.construct(this, [pattern, flags])`); maybe with `isConstructor(this) ? this : RegExp`.
My hint at `Reflect` was unnecessary, I mistakenly thought that we could use it for variadic arguments - which are not needed here.

**Comment by ljharb** [comment]
yeah basically `new this` in the implementation of `make` should work fine, presuming that it includes a spec step that throws if it doesn't have a `[[RegExpMatcher]]` internal slot.

**Comment by bergus** [comment]
Nah, there is no `[[RegExpMatcher]]` slot on regex _constructor functions_, it's an instance slot.
I don't think it's possible to detect whether a function is an ES6 sub`class` that `extends RegExp` and calls `super` to it, but I don't think it's necessary either. There's nothing wrong with being lax about what the constructor does - it's not our business.
The closest you could get afaik is `RegExp == this || RegExp.isPrototypeOf(this)`, but even that can be fooled.

**Comment by ljharb** [comment]
ah, good call

**Comment by mikesamuel** [comment]
Pull request: https://github.com/mikesamuel/regexp-make-js/pull/18
